### PR TITLE
IGNITE-19213 ODBC Data fetching

### PR DESCRIPTION
**Pull request by isapego**
- Ported fetching-related code from Ignite 2 (result_page, etc);
- Moved `primitive` class and some methods related to its handling to `common` and `protocol` libraries;
- Ported tests from Ignite 2. Fixed found bugs;
- Added some new tests.

### CODE REVIEW

File: modules/platforms/cpp/ignite/odbc/query/data_query.h [hunk 1]
```diff
@@ -129,7 +129,27 @@ class data_query : public query
      *
      * @return true, if all cursors closed remotely.
      */
-    [[nodiscard]] bool is_closed_remotely() const;
+    [[nodiscard]] bool is_closed_remotely() const {
+        return !m_has_more_pages;
```
Comment (ademakov) [inline modules/platforms/cpp/ignite/odbc/query/data_query.h:133]: In the presence of `has_more_pages()` method this method seems a bit odd.

File: modules/platforms/cpp/ignite/odbc/query/cursor.h [hunk 2]
```diff
@@ -35,41 +35,98 @@ class cursor
...
 
     /**
      * Check if the cursor has data.
      *
      * @return True if the cursor has data.
      */
-    [[nodiscard]] bool has_data() const {
-        return false;
+    [[nodiscard]] bool has_data() const
+    {
```
Comment (ademakov) [inline modules/platforms/cpp/ignite/odbc/query/cursor.h:82]: Here and above (in the `next()` method) formatting does not match our coding style.

File: modules/platforms/cpp/ignite/odbc/app/application_data_buffer.cpp
Comment (ademakov) [inline modules/platforms/cpp/ignite/odbc/app/application_data_buffer.cpp:1]: The whole file needs to be formatted.

**Review by ademakov — APPROVED** [review]

File: modules/platforms/cpp/ignite/odbc/query/data_query.h (hunk 1 above)
Comment (isapego) [reply modules/platforms/cpp/ignite/odbc/query/data_query.h:133]: Agree. This is ported code so yeah, may look a bit weird. Fixed.

File: modules/platforms/cpp/ignite/odbc/query/cursor.h (hunk 2 above)
Comment (isapego) [reply modules/platforms/cpp/ignite/odbc/query/cursor.h:82]: Thanks. Run clang-format.

File: modules/platforms/cpp/ignite/odbc/app/application_data_buffer.cpp
Comment (isapego) [reply modules/platforms/cpp/ignite/odbc/app/application_data_buffer.cpp:1]: Done

**Pull request merged**